[torchgen] Move Executorch unboxing logic into torchgen

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89596
* #90806
* #90099
* __->__ #90098

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `ExecutorchCppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D42162582](https://our.internmc.facebook.com/intern/diff/D42162582)

larryliu0820 committed:
[torchgen] Move Executorch unboxing logic into torchgen



Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

ezyang commented on torchgen/executorch/api/unboxing.py:
```diff
@@ -0,0 +1,224 @@
+from dataclasses import dataclass
+from typing import Callable, List, Sequence, Tuple, Union
```
What's the relationship between this and torchgen.api.unboxing? Can we delete torchgen.api.unboxing now?

larryliu0820 replied:
no, `torchgen.api.unboxing` is doing unboxing for IValue. `torchgen.executorch.api.unboxing` is doing unboxing for EValue.

ezyang commented on torchgen/executorch/api/unboxing.py:
```diff
@@ -0,0 +1,224 @@
+    Type,
+)
+
+connector = "\n\t"
```
you sure you want to do this with tabs :think:

larryliu0820 replied:
it doesn't really matter lol. what's the canonical way of doing this? whitespaces?

ezyang replied:
well, it's pretty tough to get f-strings to indent correctly, but yeah I usually just do spaces

ezyang commented on torchgen/executorch/api/unboxing.py:
```diff
@@ -0,0 +1,224 @@
+        torch::executor::Tensor & out_base = out.to<torch::executor::Tensor>();
+
+        EXECUTORCH_SCOPE_PROF("native_call_mul.out");
+        torch::executor::mul_outf(self_base, other_base, out_base);
```
This is not really the subject of the PR, but are you not responsible for popping the values off the stack?

ezyang commented on torchgen/executorch/api/unboxing.py:
```diff
@@ -0,0 +1,224 @@
+            t.elem, elem_name
+        )
+
+        # pyre-fixme[16]: `Enum` has no attribute `Tensor`.
```
How come pyre doesn't understand 🤔

larryliu0820 replied:
copied from fbcode, can remove.

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597383).

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


## Adding the "unboxing" logic for Executorch

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `CppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597383/)!

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383)

[ghstack-poisoned]

ezyang reviewed (approved):
I am deeply suspicious that there should be a way to use translate here, but this is fine.

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597383).

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `ExecutorchCppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D41597383](https://our.internmc.facebook.com/intern/diff/D41597383)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch unboxing logic into torchgen"


This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `ExecutorchCppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.


[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42162582).
